### PR TITLE
Feature/improved vhook

### LINKED ISSUE #10: VHook
# VHook

Don't get me wrong, it's a good tool. But for now it's lacking a few important functionalities and produces quite a bit overhead. 
Would be nice if you could improve on it. If you don't have time I could do it as well. 

Below I listed what I currently miss.

## Overhead

```dart
// Would be nice to combine expect and awaitValue methods
await hook.expect(equals(*), timeout: Duration(years: 42));

// For now we have to use a two-liner for this
await hook.awaitValue(Duration(years: 42));
hook.expect(equals(*));
```

## Functionality

- Have to use `bool?` if I only want to distinguish between two states. 
- No replacement for `Completer.completeError`
- If my value can be `null` I cannot await it
- I'd like to await based on the value and not be limited to `null` checks. For example `awaitValue(Duration(years: 42), (value) => condition)`

_Originally posted by @HenrikThoroe in https://github.com/Coronon/flyde/issues/9#issuecomment-914277597_

### PR DESCRIPTION
**Pull request by Coronon**
**Merge into _feature/build-server_**
**Closes #10** 

Rebuild `VHook` into a much more useful testing tool:

## No more need for optional

Always needing to use optional can be a pain, so we remove that requirement with this PR. You can now also use `null` like any other value as it does not represent an 'empty' `VHook` anymore.

```dart 
// VHook([dynamic = default])
final hook = VHook<int>();
```

## Completer like API including method to fail with an error

Would be more convenient for new devs as well. The `completeError` method would add value to `VHook`

```dart 
// void completeValue(T value)
hook.completeValue(0);

// void completeError(Exception exception) | or dynamic exception?
hook.completeError('Unfortunately 0 is not a number today. Try again tomorrow.');
```

## Method to set / modify the value without completing

Sometimes the value is important to really call the hook completed. Like when tracking active connections or any other counter.

```dart
// T update(T Function(T?));
hook.update((int? current) => current + 1 ?? 0);

// Future<T> updateAsync(Future<T> Function(T?));
hook.updateAsync((int? current) async => current + 1 ?? 0);

// void set(T);
hook.set(1);

// void complete();
// Completes the hook with the last value and skips possible handlers.
hook.finish();
```

## Method to await based on content

If no value handler is provided, the future completes when the first value arrives from any update method. This will throw if `complete` is called and the predicate does not evaluate to true, as it will be the final value and therefore the last chance for any value to satisfy the condition.

The method should always throw on timeout. Timeout is optional.
The return value is a Future to the awaited value.

```dart 
// Future<T> awaitValue(Duration timeout, [bool Function(T)? handler]);
hook.awaitValue(Duration(eternities: 66), (int value) => value > 2);
```

## Method to await based on completion

We sometimes only care about the final result, therefore awaiting completion would be nice.

```dart
// Future<T> awaitCompletion([Duration? timeout])
hook.awaitCompletion(timeout: Duration(seconds: 69));
```

## Combined expect and await method

This method could be used instead of `awaitValue` to instantly apply a matcher on completion or try matching on every change until succesfull (or timeout).

```dart 
// Future<T> expectAsync(Matcher matcher, {String? reason, dynamic skip, Duration? timeout});
hook.expectAsync(equals(3), timeout: Duration(eternities: 66));
```

### CODE REVIEW

**Comment by Coronon** [comment]
### The first review iteration should only focus on the interface (API) as the tests will need to be migrated accordingly.

File: test/helpers/value_hook.dart [hunk 1]
```diff
@@ -1,42 +1,198 @@
...
 /// // Test that the handler was called
-/// // This will throw an error if the handler was not called.
-/// calledHandler.expect(equals(true));
+/// // If the handler was not called after 5s, throw an error
+/// calledHandler.awaitCompletion(Duration(seconds: 5));
 /// ```
 class VHook<T> {
   /// The value of the variable.
-  T _value;
+  dynamic _value;
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: I'm not a big fan of using `dynamic` for this. I'd recommend using `late T _value` and an additional `_hasValue` flag. You could add a getter and setter to throw a custom error when accessing `_value` when `_hasValue` is false. Maybe even a small utility class

File: test/helpers/value_hook.dart [hunk 2]
```diff
@@ -1,42 +1,198 @@
...
+
+  /// The error that might be completed with
+  Object _error = _VHookValue.none;
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: Why not using an optional here? Would make no sense to have an error object of type `Null`. Is `null` even derived from `Object`?

File: test/helpers/value_hook.dart [hunk 3]
```diff
@@ -1,42 +1,198 @@
...
+    _completer.complete(_value);
+  }
+
+  /// Complete with an error or exception
+  void completeError(Object error) {
+    // Can not complete multiple times
+    _assertNotCompleted();
+
+    // Save error
+    _error = error;
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart:91]: Why do u save the error?

File: test/helpers/value_hook.dart [hunk 4]
```diff
@@ -1,42 +1,198 @@
...
+  T update(T Function(T) updater, {dynamic orElse = _VHookValue.none}) {
+    // Do not change value if already completed
+    _assertNotCompleted();
+
+    if (_value != _VHookValue.none) {
+      // Some value already set -> type guarantee -> can use updater
+      set(updater(_value));
+    } else if (orElse != _VHookValue.none) {
+      // No value set + fallback provided -> use fallback
+      set(orElse);
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: because `orElse` is dynamic u could call `set` with a value not of type `T` which would assign `_value` with a type unlike `T`. I guess it would lead to some runtime type error at some point. As already pointed out in the first review comment I'd like to migrate from `dynamic` to something type safe

File: test/helpers/value_hook.dart [hunk 5]
```diff
@@ -1,42 +1,198 @@
...
+      // Some value already set -> type guarantee -> can use updater
+      set(await updater(_value));
+    } else if (orElse != _VHookValue.none) {
+      // No value set + fallback provided -> use fallback
+      // Handle both value and Future fallbacks
+      if (orElse is Future) {
+        set(await orElse);
+      } else {
+        set(orElse);
+      }
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: That's again very unsafe. Please consider migrating

**Review by HenrikThoroe — CHANGES_REQUESTED** [review]
I'd really like if you could improve upon type safety. 
This review only includes remarks on this topic, I'll add another one when you addressed it. 

The rest looks pretty good 👍

File: test/helpers/value_hook.dart (hunk 1 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: We _have_ to use dynamic here. The default is `_VHookValue.none` which internally represents a `VHook` without any value (this allows the value `none`). So this can assume any value of type `T` or `_VHookValue.none` which means it has to be dynamic. The getter does check the value though and will always return `T` which makes this an internal dynamic only.

File: test/helpers/value_hook.dart (hunk 2 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: I dont thinck that it is derived from Object, i'll look into changing this to `Object?`

File: test/helpers/value_hook.dart (hunk 3 above)
Comment (Coronon) [reply test/helpers/value_hook.dart:91]: For better user experience. You can always use call `VHook.error` to access the `Object` used in `hook.completeError(...)`.

File: test/helpers/value_hook.dart (hunk 4 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: This would throw a runtime error when calling `set(orElse)` as `set` takes one argument of type `T`. We have to keep `orElse` dynamic as we use the default of `_VHookValue.none`. This dynamic is the price we pay for making `null` a first class citizen.

File: test/helpers/value_hook.dart (hunk 5 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: The `set` still blocks unwanted values. The `is Future` is only used to determine wether we need to await the value.

**Comment by Coronon** [comment]
### TL;DR:
_Internal_ typesafety is the prize we pay for making `null` a first class citizen. The external API is still completely typesafe, as long as you specify a type when creating a `VHook`:

```dart
final hook = VHook<int>();   // <--- Externally completely typesafe
```

```dart
final hook = VHook();   // <--- Not typesafe as <dynamic> is assumed by Dart
```

We allow this unsafe '\<dynamic>' to avoid having to use whatever type for simple 'didCall' checking, which now only requires `awaitComplete` and `complete`.

The real 'fix' here would be Union types, which dart sadly doesn't have.
You know how much I hate `dynamic`... :(

**Comment by HenrikThoroe** [comment]
You don't **have** to use `dynamic`. There are ways around this, like already discussed. If you want me to fix the type unsafety, you can make the call. And as you already mentioned yourself, the API isn't typesafe either (referring to `orElse` in `update`). 

Here's just a blueprint of a possible helper class to use instead of dynamic. You can test it in dartpad 
```dart
class Maybe<T> {
  late T _value;
  bool _hasValue;
  
  Maybe(this._value) : _hasValue = true;
  
  Maybe.empty() : _hasValue = false;
  
  set value(T val) {
    _value = val;
    _hasValue = true;
  }
  
  T get value {
    if (!_hasValue) {
      throw StateError('No Value');
    }
    
    return _value;
  }
  
  bool get hasValue => _hasValue;
}

void main() async {
  final int? val = null;
  final maybe = Maybe(val);
  final maybe2 = Maybe<int>.empty();
  
  
  print(maybe.value);
  print(maybe.hasValue);
  
  print(maybe2.hasValue);
  
  try {
    print(maybe2.value);
  } catch(e) {
    print(e);
  }
  
  maybe2.value = 9;
  print(maybe2.value);
}
```

File: test/helpers/value_hook.dart [hunk 6]
```diff
@@ -49,48 +223,190 @@ class VHook<T> {
...
+
+              return true;
+            } on test.TestFailure {
+              return false;
+            }
+          },
+        );
+      } on TimeoutException {
+        // We catch TimeoutException to show the results of
+        // matching the most current value (the [TestFailure])
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: Should this block be empty?

File: test/helpers/value_hook.dart [hunk 7]
```diff
@@ -49,48 +223,190 @@ class VHook<T> {
...
+  void _assertNotCompletedError() {
+    if (_error != null) throw _error!;
+  }
+}
+
+/// Used to encapsulate a value.
+///
+/// This is useful to distinguish  between the value `null`
+/// and 'argument not specified' for optional arguments.
+class VHookValue<T> {
```
Comment (HenrikThoroe) [inline test/helpers/value_hook.dart]: The name does not really transport much information imo. 
I'd recommend a more general type like `[Value]Container` or `[Value]Wrapper`

**Review by HenrikThoroe — CHANGES_REQUESTED** [review]
If u update the update all the tests, we might be able to complete the current two PRs within at maximum quarter a century! I know we are amazingly fast but stay calm otherwise we loose control of our work :P

File: test/helpers/value_hook.dart (hunk 6 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: Yes, but it is inconsistent with the docs. We have two options here:

1. After timeout throw a `TimeoutException`
2. After timeout use `test.expect`, which we know will throw (as we would have returned before if it wouldn’t) to show the user that expected and current state (which I would consider more useful)

The doc comment currently suggests 1. but the actual behaviour is 2..

File: test/helpers/value_hook.dart (hunk 7 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: I wanted to keep the name close to its only use case, but can of course rename this.

Comment (HenrikThoroe) [reply test/helpers/value_hook.dart]: Maybe just add a `Container` to the name and we are both fine :)

File: test/helpers/value_hook.dart (hunk 6 above)
Comment (HenrikThoroe) [reply test/helpers/value_hook.dart]: I'd throw a timeout exception on timeout. You know the keep things simple stuff ;)

File: test/helpers/value_hook.dart (hunk 7 above)
Comment (Coronon) [reply test/helpers/value_hook.dart]: `VHookValueContainer` is way to long in my opinion, let's just use `ValueContainer` then

**Review by HenrikThoroe — APPROVED** [review]
Great job migrating all the tests. Props on you for even utilizing the new features instead of an "syntax-only" migration. 

The tests are all succeeding and they fail as expected when breaking some library code.

**Comment by Coronon** [comment]
Thank you! I thought, that we should demonstrate the strengths of the new implementation ;)
